### ISSUE TITLE
Source hangs after two polls

### ISSUE THREAD

**Issue by jyates** [akka/alpakka-kafka#1068]
### Versions used 

Scala 2.11.7
Java 1.8.0_181
Akka version: 2.5.29
alpakka-kafka verison: 2.0.2
Kafka client: 2.4.0 and 5.4.0-ccs (confluent equivalent)
embedded-kafka: 2.4.0

### Expected Behavior

Poll all the records, even when the max.poll.records is small.

### Actual Behavior

Polls the records, but doesn't seem to be properly adding them to the buffer in the BaseSingleSourceLogic when recursing into `pump()` when I step through with a debugger. There does not appear to be a rebalance or anything untoward happening during the test.

Previous version that this test case (modulo minor changes like serializer and the draining control) worked with older versions:

alpakka-kafka verison: 1.0.4
Kafka client: 2.1.1-cp1
scalatest-embedded-kafka: 0.16.0

### Reproducible Test Case

This also does similar things with `max.poll.records = 1`, but this also makes the point that is still just two polls - each polling two records - that add records to the source.

Test also checks that there are indeed more than the returned records in the topic.

```
import java.util.UUID

import akka.Done
import akka.actor.ActorSystem
import akka.kafka.scaladsl.Consumer
import akka.kafka.{ConsumerSettings, Subscriptions}
import akka.stream.scaladsl.Keep
import akka.stream.testkit.scaladsl.TestSink
import akka.stream.{ActorMaterializer, ActorMaterializerSettings}
import akka.testkit.TestKit
import com.typesafe.config.ConfigFactory
import net.manub.embeddedkafka.{Codecs, EmbeddedKafka, EmbeddedKafkaConfig}
import org.apache.kafka.common.serialization.Deserializer
import org.junit.runner.RunWith
import org.scalatest.Matchers._
import org.scalatest.concurrent.ScalaFutures
import org.scalatest.junit.JUnitRunner
import org.scalatest.{BeforeAndAfterAll, BeforeAndAfterEach, PartialFunctionValues, WordSpecLike}
import org.slf4j.{Logger, LoggerFactory}

import scala.concurrent.duration._

@RunWith(classOf[JUnitRunner])
final class KafkaStreamsPollIntegrationSpec
    extends TestKit(ActorSystem("KafkaStreamsPollIntegrationSpec", ConfigFactory.load("testing")))
    with WordSpecLike
    with PartialFunctionValues
    with BeforeAndAfterAll
    with BeforeAndAfterEach
    with EmbeddedKafka
    with ScalaFutures {

  private implicit val logger: Logger = LoggerFactory.getLogger(classOf[KafkaStreamsPollIntegrationSpec])
  private val materializerSettings = ActorMaterializerSettings(system)
  private implicit val materializer = ActorMaterializer(materializerSettings)
  private implicit val ec = system.dispatcher

  private val kafkaConfig = EmbeddedKafkaConfig(kafkaPort = 0, zooKeeperPort = 0)
  private val serializer = Codecs.stringSerializer
  private val deserializer = Codecs.stringDeserializer

  "Alpakka kafka" should {
    "poll more than twice" in {
      withRunningKafkaOnFoundPort(kafkaConfig) { implicit actualKafkaConfig: EmbeddedKafkaConfig =>
        val topic = "topic"
        // write some simple string data
        val records = (1 to 10).map(x => (s"key-$x", s"value-$x"))
        publishToKafka(topic, records)(config = actualKafkaConfig, keySerializer = serializer, serializer = serializer)
        // make sure we are not crazy and the data is present
        implicit val deserialize = deserializer.asInstanceOf[Deserializer[AnyRef]]
        consumeNumberMessagesFrom(topic, 10) should have size 10

        val consumerConf = system.settings.config.getConfig("akka.kafka.consumer")
        val settings = ConsumerSettings(consumerConf, deserializer, deserializer)
            .withBootstrapServers(s"localhost:${actualKafkaConfig.kafkaPort}")
            .withProperty("auto.offset.reset", "earliest")
            .withProperty("max.poll.records", "2")
            .withGroupId(s"foo-bar-${UUID.randomUUID()}")

        val subscription = Subscriptions.topics(topic)
        val source = Consumer.plainSource(settings, subscription)

        val (control, probe) = source
            .log("peek")
            .map(_.value)
            .toMat(TestSink.probe)(Keep.both)
            .run()
        probe.request(10)
        for (i <- Range(1, 11)) {
          println(s"Expecting $i ....")
          println(s"    -> ${probe.expectNext(10.seconds)}")
        }
        probe.cancel()
        control.isShutdown.futureValue shouldBe Done
      }
    }
  }
}
```

### Relevant logs

Running the test with debugging turned up, you see that it does request the records, but they never end up getting sent downstream.
```
Expecting 1 ....
2020-02-29 02:12:47 INFO  a.kafka.internal.SingleSourceLogic[92] - [57d7c] Starting. StageActor Actor[akka://KafkaStreamsPollIntegrationSpec/system/StreamSupervisor-0/$$a#-1314779994]
2020-02-29 02:12:47 DEBUG a.kafka.internal.KafkaConsumerActor[97] - [5ec16] Starting Actor[akka://KafkaStreamsPollIntegrationSpec/system/kafka-consumer-1#-481845586]
2020-02-29 02:12:47 DEBUG a.kafka.internal.KafkaConsumerActor[97] - [5ec16] Creating Kafka consumer with akka.kafka.ConsumerSettings(properties=auto.offset.reset -> earliest,group.id -> foo-bar-29a5328b-d373-4e23-bc8f-fcd9257e609b,bootstrap.servers -> localhost:65047,enable.auto.commit -> false,max.poll.records -> 2,keyDeserializer=Some(org.apache.kafka.common.serialization.StringDeserializer@612100d2),valueDeserializer=Some(org.apache.kafka.common.serialization.StringDeserializer@612100d2),pollInterval=50 milliseconds,pollTimeout=50 milliseconds,stopTimeout=30 seconds,closeTimeout=20 seconds,commitTimeout=15 seconds,commitRefreshInterval=Duration.Inf,dispatcher=akka.kafka.default-dispatcher,commitTimeWarning=1 second,waitClosePartition=500 milliseconds,metadataRequestTimeout=5 seconds,drainingCheckInterval=30 milliseconds,connectionCheckerSettings=akka.kafka.ConnectionCheckerSettings(enable=false,maxRetries=3,checkInterval=15 seconds,factor=2.0),partitionHandlerWarning=5 secondsenrichAsync=None)
2020-02-29 02:12:47 DEBUG a.kafka.internal.KafkaConsumerActor[97] - received handled message Subscribe(Set(topic),Chain(Chain(Chain(EmptyPartitionAssignmentHandler, akka.kafka.internal.SingleSourceLogic$$anon$2@70c6f3e9), AsyncCallbacks(TopicSubscription(Set(topic),None,EmptyPartitionAssignmentHandler), Actor[akka://KafkaStreamsPollIntegrationSpec/system/StreamSupervisor-0/$$a#-1314779994])), akka.kafka.internal.SingleSourceLogic$$anon$2@5d1537c4)) from Actor[akka://KafkaStreamsPollIntegrationSpec/system/StreamSupervisor-0/$$a#-1314779994]
2020-02-29 02:12:47 DEBUG a.kafka.internal.KafkaConsumerActor[97] - received handled message Poll(akka.kafka.internal.KafkaConsumerActor@4b805292,true) from Actor[akka://KafkaStreamsPollIntegrationSpec/deadLetters]
2020-02-29 02:12:47 DEBUG a.kafka.internal.KafkaConsumerActor[97] - received handled message Poll(akka.kafka.internal.KafkaConsumerActor@4b805292,true) from Actor[akka://KafkaStreamsPollIntegrationSpec/deadLetters]
2020-02-29 02:12:47 DEBUG a.kafka.internal.KafkaConsumerActor[97] - received handled message Poll(akka.kafka.internal.KafkaConsumerActor@4b805292,true) from Actor[akka://KafkaStreamsPollIntegrationSpec/deadLetters]
2020-02-29 02:12:47 DEBUG a.kafka.internal.KafkaConsumerActor[97] - received handled message Poll(akka.kafka.internal.KafkaConsumerActor@4b805292,true) from 

... (bunch of the dead-letter handled polls) ...

2020-02-29 02:20:20 DEBUG a.kafka.internal.KafkaConsumerActor[97] - received handled message Poll(akka.kafka.internal.KafkaConsumerActor@81ea120,true) from Actor[akka://KafkaStreamsPollIntegrationSpec/deadLetters]
2020-02-29 02:20:20 DEBUG a.kafka.internal.SingleSourceLogic[97] - [3d1d4] Assigned partitions: Set(topic-0). All partitions: Set(topic-0)
2020-02-29 02:20:20 DEBUG a.kafka.internal.SingleSourceLogic[97] - [3d1d4] Requesting messages, requestId: 1, partitions: Set(topic-0)
2020-02-29 02:20:20 DEBUG a.kafka.internal.KafkaConsumerActor[97] - received handled message RequestMessages(1,Set(topic-0)) from Actor[akka://KafkaStreamsPollIntegrationSpec/system/StreamSupervisor-0/$$a#347326929]
    -> value-1
Expecting 2 ....
    -> value-2
Expecting 3 ....
2020-02-29 02:20:20 DEBUG a.kafka.internal.SingleSourceLogic[97] - [3d1d4] Requesting messages, requestId: 2, partitions: Set(topic-0)
2020-02-29 02:20:20 DEBUG a.kafka.internal.KafkaConsumerActor[97] - received handled message RequestMessages(2,Set(topic-0)) from Actor[akka://KafkaStreamsPollIntegrationSpec/system/StreamSupervisor-0/$$a#347326929]
    -> value-3
Expecting 4 ....
    -> value-4
Expecting 5 ....
2020-02-29 02:20:20 DEBUG a.kafka.internal.SingleSourceLogic[97] - [3d1d4] Requesting messages, requestId: 3, partitions: Set(topic-0)
2020-02-29 02:20:20 DEBUG a.kafka.internal.KafkaConsumerActor[97] - received handled message RequestMessages(3,Set(topic-0)) from Actor[akka://KafkaStreamsPollIntegrationSpec/system/StreamSupervisor-0/$$a#347326929]
2020-02-29 02:20:20 DEBUG a.kafka.internal.SingleSourceLogic[97] - [3d1d4] Requesting messages, requestId: 4, partitions: Set(topic-0)
2020-02-29 02:20:20 DEBUG a.kafka.internal.KafkaConsumerActor[97] - received handled message RequestMessages(4,Set(topic-0)) from Actor[akka://KafkaStreamsPollIntegrationSpec/system/StreamSupervisor-0/$$a#347326929]
2020-02-29 02:20:20 DEBUG a.kafka.internal.SingleSourceLogic[97] - [3d1d4] Requesting messages, requestId: 5, partitions: Set(topic-0)
2020-02-29 02:20:20 DEBUG a.kafka.internal.KafkaConsumerActor[97] - received handled message RequestMessages(5,Set(topic-0)) from Actor[akka://KafkaStreamsPollIntegrationSpec/system/StreamSupervisor-0/$$a#347326929]
2020-02-29 02:20:20 DEBUG a.kafka.internal.SingleSourceLogic[97] - [3d1d4] Requesting messages, requestId: 6, partitions: Set(topic-0)
2020-02-29 02:20:20 DEBUG a.kafka.internal.KafkaConsumerActor[97] - received handled message RequestMessages(6,Set(topic-0)) from Actor[akka://KafkaStreamsPollIntegrationSpec/system/StreamSupervisor-0/$$a#347326929]
2020-02-29 02:20:20 DEBUG a.kafka.internal.KafkaConsumerActor[97] - received handled message Poll(akka.kafka.internal.KafkaConsumerActor@81ea120,true) from Actor[akka://KafkaStreamsPollIntegrationSpec/deadLetters]

... (bunch of the dead-letter handled polls)...

2020-02-29 02:12:51 DEBUG a.kafka.internal.KafkaConsumerActor[97] - received handled message Poll(akka.kafka.internal.KafkaConsumerActor@4b805292,true) from Actor[akka://KafkaStreamsPollIntegrationSpec/deadLetters]
... more dead-letter polls ...
- should poll more than twice *** FAILED *** (25 seconds, 222 milliseconds)
  java.lang.AssertionError: Expected OnNext(_), yet no element signaled during 10 seconds
```

**Comment by seglo** [comment]
Hi @jyates. I tried adding your test to the `tests` project and running it with Scala 2.11.7.  Initially I had some issues actually compiling the project because of a compiler regression related to accessibility of case class private constructors that goes away in 2.11.8 (may have been this: https://github.com/scala/bug/issues/9546). I worked around the problem (note the package private constructor modifications to the `Result` types in the test branch PR I published) to run the test and was able to recreate your issue.  After I got the project to build I was able to recreate your issue.  Since the version of 2.11 you're running seemed very old (released June 23, 2015), I incrementally tried later versions and found that the issue was no longer reproducible in 2.11.9. It appears to be caused by another Scala regression where appending to an empty iterator results in a new iterator that still reports as empty (`hasNext == false`), as reported in this Scala issue: https://github.com/scala/bug/issues/9766 and resolved in 2.11.9. Keep in mind that 2.11.9 and 2.11.10 shouldn't be used for other reasons.  Users are advised to upgrade from 2.11.8 to at least 2.11.11.

I would suggest that you at least upgrade to 2.11.12 (the latest 2.11 revision), or if possible directly to 2.13.2.  We recommend that users always use at least the version of Scala that was used to compile the artifact.  For Alpakka Kafka 2.0.2 this is Scala 2.11.12.

Here's the draft PR with your test, my compiler workarounds, and notes on reproducibility.  https://github.com/akka/alpakka-kafka/pull/1072
